### PR TITLE
publish docs at blueskyproject.io

### LINKED ISSUE #59: add GH Actions workflow to publish the docs
> Ah, github.io, not RTD. The docs are available here: https://blueskyproject.io/hklpy/master/
> 
> Is there a workflow or batch file to be examined?

I think we can reuse https://github.com/bluesky/ophyd/blob/master/.github/workflows/publish-docs.yml for the docs. Is `.travis.yml` fully retired for now?

_Originally posted by @mrakitin in https://github.com/bluesky/hklpy/issues/52#issuecomment-742787007_

### PR DESCRIPTION
**Pull request by prjemian**
fixes #59

https://blueskyproject.io/hklpy/master

always pushes to *master* branch of blueskyproject.io

### CODE REVIEW

**Comment by mrakitin** [comment]
For some reason it complains about a missing `jupyter`:
```
make: Entering directory '/home/runner/work/hklpy/hklpy/examples'
jupyter nbconvert --output-dir ../docs/source/examples/notebooks --to rst e6c.ipynb
make: jupyter: Command not found
make: *** [e6c.rst] Error 127
Makefile:21: recipe for target 'e6c.rst' failed
make: Leaving directory '/home/runner/work/hklpy/hklpy/examples'
Error: Process completed with exit code 2.
```

Was the env activated?

**Comment by prjemian** [comment]
That's what is missing.  It *should* be activated: https://github.com/bluesky/hklpy/blob/8d236cfd41b3af7ba735b965dde6e7145ef067b5/.github/workflows/publish-docs.yml#L32-L34

**Comment by mrakitin** [comment]
> That's what is missing. It _should_ be activated:
> 
> https://github.com/bluesky/hklpy/blob/8d236cfd41b3af7ba735b965dde6e7145ef067b5/.github/workflows/publish-docs.yml#L32-L34

It also should be activated in the step where you make the docs, I believe. I had a similar issue with Azure Pipelines.

**Comment by mrakitin** [comment]
I may be wrong. I was thinking about separate jobs. The env should be preserved between steps.

**Comment by prjemian** [comment]
Next step is to insert tmate into the flow

**Comment by prjemian** [comment]
@mrakitin muttering, just pushed docs to ophyd, confessed on slack, then set it right again

On the positive side, you can see a version of the built documentation for the *24-examples* branch now at https://blueskyproject.io/hklpy

**Comment by prjemian** [comment]
Ready for review

File: .github/workflows/publish-docs.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        channel-priority: true
+        channels: defaults,conda-forge
+        environment-file: environment.yml
+        mamba-version: "*"
+        python-version: ${{ matrix.python-version }}
+        use-only-tar-bz2: true  # required for caching
+
+    - name: Install publishing requirements
+      run: |
+        source ${CONDA}/bin/activate activate ${ENV_NAME}
```
Comment (mrakitin) [inline .github/workflows/publish-docs.yml:48]: Does not look like we need the second `activate`.

Would the newer approach work?
```py
source ${CONDA}/etc/profile.d/conda.sh
conda activate ${ENV_NAME}
```

Comment (prjemian) [reply .github/workflows/publish-docs.yml:48]: I've added comparison tests to the workflow, show `conda info` result as test for each.

activation method | active environment
--- | ---
default | [hklpy-docs-3.8](https://github.com/bluesky/hklpy/runs/1537271700?check_suite_focus=true#step:7:9)
standard activate | [hklpy-docs-3.8](https://github.com/bluesky/hklpy/runs/1537271700?check_suite_focus=true#step:8:10)
newer activate | [hklpy-docs-3.8](https://github.com/bluesky/hklpy/runs/1537271700?check_suite_focus=true#step:9:12)

Which means it should not be necessary to activate explicitly.  This instruction should activate for us: https://github.com/bluesky/hklpy/blob/4187261b74ce5c7f73530e1710891b5740642354/.github/workflows/publish-docs.yml#L35-L37

So why was it not working previously? https://github.com/bluesky/hklpy/runs/1533866545?check_suite_focus=true#step:8:10  That's why the explicit activation was added in commit 108674f527a0e73be6b715cb868323f68f0b2f16.

Perhaps it was the simultaneous addition of explicit install environment here that resolved the `jupyter not found` : https://github.com/bluesky/hklpy/blob/108674f527a0e73be6b715cb868323f68f0b2f16/.github/workflows/publish-docs.yml#L45

Comment (prjemian) [reply .github/workflows/publish-docs.yml:48]: Also note syntax error in the activate commands of previous scripts: https://github.com/bluesky/hklpy/blob/4187261b74ce5c7f73530e1710891b5740642354/.github/workflows/publish-docs.yml#L78

Too many `activate` words.  Benign.  Does not seemed to have caused an error.

**Comment by prjemian** [comment]
And, with no explicit conda activation, publishing errors with [`make: jupyter: Command not found`](https://github.com/bluesky/hklpy/runs/1537412718?check_suite_focus=true#step:7:47) because there is *no* conda environment activated: [`active environment : None`](https://github.com/bluesky/hklpy/runs/1537412718?check_suite_focus=true#step:7:16)

**Comment by prjemian** [comment]
Publishing requirements must be installed into named environment, too. [`   conda install jupyter nbconvert sphinx sphinxcontrib-napoleon -c defaults -c conda-forge`](https://github.com/bluesky/hklpy/runs/1537412718?check_suite_focus=true#step:5:3)

**Comment by prjemian** [comment]
Possible related to this early warning: [` Warning: "channels" set on the "environment-file" do not match "channels" set on the action!`](https://github.com/bluesky/hklpy/runs/1537524894?check_suite_focus=true#step:4:744)

**Comment by prjemian** [comment]
Confirmed that `environment.yml` file updated by `sed` [according to our intentions](https://github.com/bluesky/hklpy/runs/1537577722?check_suite_focus=true#step:3:9) and backup file created with supplied environment name.

**Comment by prjemian** [comment]
Success with 4d28043 yet still same warning about inconsistent environment name specification.  Testing for possible indentation error.

**Comment by prjemian** [comment]
Seems to be an important lesson with the [`conda-incubator/setup-miniconda` action](https://github.com/conda-incubator/setup-miniconda/tree/8246c3d5d22bc5ca3416ddf4f74b77c2fb1d33d0#IMPORTANT):

> Bash shells do not use `~/.profile` or `~/.bashrc` so these shells need to be explicitely declared ...

Does not explain the early warning about inconsistent names.

**Comment by prjemian** [comment]
Looking at the [source code for that action](https://github.com/conda-incubator/setup-miniconda/blob/8246c3d5d22bc5ca3416ddf4f74b77c2fb1d33d0/action.yml#L40):

```
  activate-environment:
    description:
      'Environment to activate by default on all shells. Default is "test". If
      an empty string is used, no environment is activated by default (For
      "base" activation see the "auto-activate-base" option). If the environment
      does not exist, it will be created and activated. If "environment-file" is
      used and you want that to be the environment used, you need to explicitely
      provide the name of that environmet on "activate-environment". If using
      sh/bash/cmd.exe shells please read the IMPORTANT! section on the
      README.md! to properly activate conda environments on these shells.'
    required: false
    default: "test"
```

Let's switch the `sed` command to set the environment name to the default `test` and eliminate the `activate-environment` command.

**Comment by prjemian** [comment]
Warning is gone.  Now apply the [important lesson](https://github.com/bluesky/hklpy/pull/60#issuecomment-743224251) from above.

**Comment by prjemian** [comment]
Commit 4568d4a describes the fix to the root cause why conda environment was not activated as expected.  Must always declare the shell when using `run` (second line of example) as in:

```
    - name: Install publishing requirements
      shell: bash -l {0}
      run: |
        conda install jupyter nbconvert sphinx sphinxcontrib-napoleon -c defaults -c conda-forge
        pip install sphinx-rtd-theme
```

**Comment by prjemian** [comment]
Address this warning: [` Warning: "channels" set on the "environment-file" do not match "channels" set on the action!`](https://github.com/bluesky/hklpy/runs/1537827707?check_suite_focus=true#step:4:24) following advice from the action's [`README.md`](https://github.com/conda-incubator/setup-miniconda/tree/8246c3d5d22bc5ca3416ddf4f74b77c2fb1d33d0#IMPORTANT)

> If you plan to use a `environment.yaml` file to set up the environment, the action will read the `channels` listed in the key (if found). If you provide the `channels` input in the action they must not conflict with what was defined in `environment.yaml`, otherwise the conda solver might find conflicts and result in very long install times.

**Comment by prjemian** [comment]
Hmm. [`Error: "mamba-version=*" requires "conda-forge" to be included in "channels!"`](https://github.com/bluesky/hklpy/runs/1537910420?check_suite_focus=true#step:4:17)

**Comment by prjemian** [comment]
But warning still present after cf49b52.  Another benign non-problem.  Ignore it.

**Comment by prjemian** [comment]
All is good now, configure for normal production publishing (only pushes to main branch).

**Comment by prjemian** [comment]
@mrakitin Ready for review again.

**Comment by danielballan** [comment]
FYI I am working on a way to avoid hard-coding the repo name in the `publish_docs.yml` workflow, removing the "trap" where copying the workflow from one repo to another can cause an overwrite push.

**Comment by danielballan** [comment]
GitHub doesn't give you access to any variable with _just_ the repo name, only `{org}/{repo}` so some bash string manipulation is required...

**Comment by prjemian** [comment]
@danielballan You can work that up that on this branch!

**Comment by prjemian** [comment]
Just swap these changes to publish from this branch.  https://github.com/bluesky/hklpy/pull/60/commits/9c5b479ff1b23d6301e42ee3454e1aa97925c0ac

**Comment by prjemian** [comment]
https://stackoverflow.com/questions/3162385/how-to-split-a-string-in-shell-and-get-the-last-field

> You can use string operators:
>
> ```
> $ foo=1:2:3:4:5
> $ echo ${foo##*:}
> 5
> ```

**Comment by danielballan** [comment]
Thanks. I found string operators but I'm flailing trying to figure how to set env vars in actions that persist between "steps". This seems to be it. https://github.com/actions/starter-workflows/issues/68#issuecomment-592997170

**Comment by prjemian** [comment]
Put the env vars in the `jobs` section:

```
jobs:
  build:
    env:
      ENV_NAME: test
      PACKAGE: looky-here-for-the-package-repo-name--put-that-magic-here

```

**Comment by danielballan** [comment]
Is "looky here..." allowed to be bash? It would be ideal to extract the name programmatically.

**Comment by danielballan** [comment]
Not sure why it doesn't like this: https://github.com/bluesky-sandbox/ophyd/actions/runs/415721222/workflow

**Comment by prjemian** [comment]
It's not clear that a shell is available when `env` is defined.

**Comment by danielballan** [comment]
Ah ha! I was missing a space on Line 32, just below the line linked. That approach may work after all...

**Comment by danielballan** [comment]
Follow along for the fun... https://github.com/bluesky-sandbox/ophyd/runs/1538168238?check_suite_focus=true

**Comment by danielballan** [comment]
Hmm. Looks like I want https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows. GH Actions is a whole new world....

**Comment by danielballan** [comment]
Flashbacks to the first project I ever got working on Travis-CI (trackpy). It took about 300 tries. :-D

**Comment by danielballan** [comment]
The push worked, but the version number in the docs is still wrong: https://github.com/bluesky-sandbox/bluesky-sandbox.github.io/blob/348895075d59b26dfca2bbd8fa068cab686756c0/ophyd/index.html#L73

**Comment by prjemian** [comment]
I'm suspicious of a version mismatch in the versioneer package as the
reason for the wrong version number formatted in the ophyd docs.

**Comment by prjemian** [comment]
Can you set a version requirement on the versioneer install?

**Comment by danielballan** [comment]
Ah, I'm testing this on `bluesky-sandbox/ophyd`, not `bluesky/ophyd`, I have to sync the tags over:

```
git push sandbox --tags
```

They were never uploaded to the mirror of ophyd at `bluesky-sandbox/ophyd`. We'll see if that does it....

**Comment by danielballan** [comment]
That didn't do it. https://github.com/bluesky-sandbox/bluesky-sandbox.github.io/blob/master/ophyd/index.html#L73

The version comes from here: https://github.com/bluesky/ophyd/blob/251b9bdf288377178302db37a876b0ad60a5e65d/docs/source/conf.py#L63

which is set by:
https://github.com/bluesky/ophyd/blob/251b9bdf288377178302db37a876b0ad60a5e65d/ophyd/__init__.py#L102

which runs:

https://github.com/bluesky/ophyd/blob/251b9bdf288377178302db37a876b0ad60a5e65d/ophyd/_version.py#L477-L520

and consults `setup.cfg` (which works for _published_ releases) and `git` (for source installations, as in this case).

**Comment by danielballan** [comment]
In installation log we can that `pip list` reports that the installation version of ophyd is `0.post1+gbd58335`.

https://github.com/bluesky-sandbox/ophyd/runs/1538323461?check_suite_focus=true#step:4:666

**Comment by danielballan** [comment]
Locally:

```
git clone https://github.com/bluesky-sandbox/ophyd
cd ophyd/
git fetch --tags
ython3 -m venv /tmp/venv
source /tmp/venv/bin/activate
pip install setuptools wheel
pip install .
```

works:

```
pip list | grep ophyd
ophyd         1.5.4.post88+gbd58335
```

Not sure what is happening differently in Actions.

File: .github/workflows/publish-docs.yml (hunk 1 above)
Comment (mrakitin) [reply .github/workflows/publish-docs.yml]: Neat! Thanks for performing the tests, so we know exactly how to properly activate the env in the future.

**Comment by mrakitin** [comment]
@prjemian, @danielballan, is it ready for another round of review? Are you happy with the versions, or does it need more work?

Does this PR supersede #52?

**Review by danielballan — APPROVED** [review]

**Comment by danielballan** [comment]
There a couple things we might add (like a scheduled CRON-like run of the tests to help detect breakages due to upstream changes early) but maybe it would be best to do that in a second pass once we've made more progress on other repos and have a standard pattern. I expect we'll keep finding tweaks and revisions. No reason to hold up this PR over that.

**Comment by prjemian** [comment]
This PR will merge with #52 in the 24-examples branch.  Once we're happy enough with that, #52 will merge 24-examples into main.

**Comment by prjemian** [comment]
@mrakitin : If you are OK with this, then please merge (into branch *24-examples*), then delete the branch if it lets you.

@mrakitin , @danielballan Thanks for taking the time on this interesting distraction.

**Review by mrakitin — APPROVED** [review]
Thanks!

**Pull request merged**